File: python/src/practice_problems/binary_tree.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.inorder_output: List[int] = []
        self.inorder(root)
        return self.inorder_output

    def inorder(self, node: Optional[TreeNode]):
        if node is None:
            return

        self.inorder(node.left)
        self.inorder_output.append(node.val)
        self.inorder(node.right)

    def inorderTraversalIterative(self, root: Optional[TreeNode]) -> List[int]:
        stack: List[TreeNode] = []
        inorder_output: List[int] = []
        current = root
        while current is not None:
            stack.append(current)
            current = current.left
        while len(stack) != 0:
            current = stack.pop()
            inorder_output.append(current.val)

            if current.right:
                current = current.right
                while current is not None:
                    stack.append(current)
                    current = current.left

        return inorder_output

File: rust/practice_problems/src/binary_tree.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Box<TreeNode>>,
    pub right: Option<Box<TreeNode>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub struct Solution;
impl Solution {
    pub fn inorder_traversal(root: Option<Box<TreeNode>>) -> Vec<i32> {
        let mut result = Vec::new();
        Self::inorder(root, &mut result);
        result
    }

    fn inorder(node: Option<Box<TreeNode>>, result: &mut Vec<i32>) {
        if let Some(node) = node {
            Self::inorder(node.left, result);
            result.push(node.val);
            Self::inorder(node.right, result);
        }
    }

    pub fn inorder_traversal_iterative(root: Option<Box<TreeNode>>) -> Vec<i32> {
        let mut result = Vec::new();

        let mut stack: Vec<TreeNode> = Vec::new();
        let mut current = root;

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let input = Some(Box::new(TreeNode {
            val: 1,
            left: Some(Box::new(TreeNode {
                val: 2,
                left: None,
                right: None,
            })),
            right: Some(Box::new(TreeNode {
                val: 3,
                left: None,
                right: None,
            })),
        }));

        let expected = vec![2, 1, 3];
        assert_eq!(Solution::inorder_traversal(input), expected);
    }
}
